mds: scrub fix

sajibreadd-croit opened the pull request:
1. remote link damage identification with reverse parent scrubbing
   - remote link identification becomes tricky if inode is cached.
   - Try to open the link normally, if issue while opening mark as damaged
   - If openned successfully, it can be possible there is damage but inode
     is cached that's why it is succssful while opening. In that case take
     that openned inode, and scrub ancestors recursively. If any of the ancestor
     is damaged it remote link is marked as damaged.
   - while scrubbing some flag is maintained in the inode,
     e.g. whether scrub is backward or forward or both
   - his backward scrubbing will only work in read-only scrub that means
     without repair flag and mds_scrub_hard_link this ceph flag is turned on.
   - A new type of damage introduced, using which multiple links point to same
     inode can be identified, which was not possible previously.
2. mds_damage_log_to_file and mds_damage_log_file is used to print out damages
   in a file persistently as it's not safe to keep it in memory

<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins test classic perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-classic/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test crimson perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-crimson/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test signed` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pr-commits/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-commits/config/definitions/ceph-pr-commits.yml)
- `jenkins test make check` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests/config/definitions/ceph-pull-requests.yml)
- `jenkins test make check arm64` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests-arm64/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests-arm64/config/definitions/ceph-pull-requests-arm64.yml)
- `jenkins test submodules` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-pr-submodules/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-submodules/config/definitions/ceph-pr-commits.yml)
- `jenkins test dashboard` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-pull-requests/config/definitions/ceph-dashboard-pull-requests.yml)
- `jenkins test dashboard cephadm` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-cephadm-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-cephadm-e2e/config/definitions/ceph-dashboard-cephadm-e2e.yml)
- `jenkins test api` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-api/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-api/config/definitions/ceph-pr-api.yml)
- `jenkins test docs` [ReadTheDocs](https://readthedocs.org/projects/ceph/) | [Github Workflow Definition](https://github.com/ceph/ceph/blob/main/.readthedocs.yml)
- `jenkins test ceph-volume all` [Jenkins Jobs](https://jenkins.ceph.com/view/ceph-volume%20PR/) | [Jenkins Jobs Definition](https://github.com/ceph/ceph-build/blob/main/ceph-volume-cephadm-prs/config/definitions/ceph-volume-pr.yml)
- `jenkins test windows` [Jenkins Job](https://jenkins.ceph.com/job/ceph-windows-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-windows-pull-requests/config/definitions/ceph-windows-pull-requests.yml)
- `jenkins test rook e2e` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-orchestrator-rook-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-rook-e2e/config/definitions/ceph-orchestrator-rook-e2e.yml)

You must only issue one Jenkins command per-comment. Jenkins does not understand
comments with more than one command.
</details>

sajibreadd committed:
qa: add a test to verify that a damage hard link id detected during scrub

Also update ignorelist with the expected cluster warning.

Fixes: https://tracker.ceph.com/issues/69345
Signed-off-by: Md Mahamudur Rahaman Sajib <mahamudur.sajib@croit.io>
(cherry picked from commit 13eb6c6ce9e6c8f5a4a8bcdc1bba64bb92393cf1)

sajibreadd-croit committed:
mds: scrub pins more inodes than the mds_cache_memory_limit
For scrubbing dirfrag we are pushing children back into the scrub stack. Instead we can follow the same
strategy for scrub directory and pushing children front of the scrub stack, and in kick_off_scrubs always
start scrubbing from the front of the stack. It will prevent ScrubStack to pinning whole level of the file-system
tree.

Fixes: https://tracker.ceph.com/issues/71167
Signed-off-by: Md Mahamudur Rahaman Sajib <mahamudur.sajib@croit.io>

sajibreadd-croit committed:
mds: gracefully terminate missed dir object scrubbing
Fixes: https://tracker.ceph.com/issues/68611
Signed-off-by: Md Mahamudur Rahaman Sajib <mahamudur.sajib@croit.io>

sajibreadd-croit committed:
mds: remote link scrub fix using backward scrubbing from head inode
1. remote link damage identification with reverse parent scrubbing
   - remote link identification becomes tricky if inode is cached.
   - Try to open the link normally, if issue while opening mark as damaged
   - If openned successfully, it can be possible there is damage but inode
     is cached that's why it is succssful while opening. In that case take
     that openned inode, and scrub ancestors recursively. If any of the ancestor
     is damaged it remote link is marked as damaged.
   - while scrubbing some flag is maintained in the inode,
     e.g. whether scrub is backward or forward or both
   - his backward scrubbing will only work in read-only scrub that means
     without repair flag and mds_scrub_hard_link this ceph flag is turned on.
   - A new type of damage introduced, using which multiple links point to same
     inode can be identified, which was not possible previously.
2. mds_damage_log_to_file and mds_damage_log_file is used to print out damages
   in a file persistently as it's not safe to keep it in memory
Signed-off-by: Md Mahamudur Rahaman Sajib <mahamudur.sajib@croit.io>

sajibreadd-croit commented on src/mds/ScrubStack.cc:
```diff
@@ -209,49 +224,53 @@ void ScrubStack::kick_off_scrubs()
+        if (added_children) {
+          // dirfrags were queued at top of stack
+          it = scrub_stack.begin();
+        }
```
@ifed01 regarding this comment https://github.com/croit/ceph/pull/5#discussion_r2632212411

> Shouldn't we call scrub_dir_inode_final(in) here if (!remote_links().empty()) ?

if (!remote_links.empty) then inside `scrub_dir_inode` it will call `scrub_dir_inode_final(in)` and handle the backward scrub there.

ifed01 replied:
hmm... But calling scrub_dir_inode_final() would happen before forward_scrub is cleared so it would be a no-op in regard to remote links, nor?

ifed01 replied:
Made a significant rework around this stuff, see (https://github.com/croit/ceph/pull/7)

sajibreadd-croit commented on src/mds/ScrubStack.cc:
```diff
@@ -408,16 +428,118 @@ class C_InodeValidated : public MDSInternalContext
+  if (fatal) {
+    mdcache->mds->damaged();
+    ceph_abort(); // unreachable, damaged() respawns us
+  }
```
@ifed01 regarding this comment https://github.com/croit/ceph/pull/5#discussion_r2632271249
> I'm not sure we should abort at this point in case of scrubbing. May be just log an error and proceed?

damages are stored in memory and there is a threshold `mds_damage_table_max_entries` is used to maintain the size of damages in the memory. It will be fatal when number of entries gets exceeded. As we will print the damages in file, `fatal` will always be false.

ifed01 replied:
I understand that but damage_table has a protection against oversized mem usage. So IIUC it just omits new entries if threashold has been reached. So why aborting MDS that's jut scrubbing then? I understand the original behavior when this happens after an attempt to open remote link but this looks redundant for scrubbing.

sajibreadd-croit replied:
It's fixed now, thanks.

ifed01 commented on src/mds/CDir.cc:
```diff
@@ -2006,7 +2007,7 @@ void CDir::_omap_fetched(bufferlist& hdrbl, map<string, bufferlist>& omap,
                      "files may be lost (" << get_path() << ")";
 
-    go_bad(complete);
+    go_bad(complete | from_scrub);
```
nit: IMO this rather implies logical OR (||) usage...

ifed01 commented on src/common/options/mds.yaml.in:
```diff
@@ -1555,3 +1555,27 @@ options:
+- name: mds_scrub_hard_link
+  type: bool
+  level: advanced
+  desc: force scrubbing hard link
```
IMO better rephrase to "force hard link scrubbing"

ifed01 commented on src/mds/DamageTable.cc:
```diff
@@ -293,6 +372,18 @@ void DamageTable::erase(damage_entry_id_t damage_id)
+    auto remote_link_entry = std::static_pointer_cast<RemoteLinkDamage>(entry);
+    auto df_remote_link_it = remote_links.find(remote_link_entry->ino);
+    if (df_remote_link_it != remote_links.end()) {
+      auto damage_it = df_remote_link_it->second.find(entry->path);
```
This and the following 3 lines could be replaced with a single erase() call:
df_remote_link_it->second.earse(entry->path);

ifed01 commented on src/mds/DamageTable.cc:
```diff
@@ -301,3 +392,57 @@ void DamageTable::erase(damage_entry_id_t damage_id)
+
+    if (!fs::exists(dir, ec)) {
+      if (ec) {
+        dout(0) << "error checking existence of damage dir: " << dir << " ("
```
use derr here and for ether errors instead?

ifed01 commented on src/mds/DamageTable.cc:
```diff
@@ -301,3 +392,57 @@ void DamageTable::erase(damage_entry_id_t damage_id)
+void DamageTable::open_damage_log_file(std::ofstream &fout,
+                                       const std::filesystem::path &file_path) {
+  namespace fs = std::filesystem;
+  log_file_opened = false;
```
One can call close_damage_log_file() here instead to save a few lines...

ifed01 commented on src/mds/MDSRank.cc:
```diff
@@ -498,7 +498,9 @@ MDSRank::MDSRank(
     timer(timer_), mdsmap(mdsmap_),
     objecter(new Objecter(g_ceph_context, msgr, monc_, ioc)),
-    damage_table(whoami_), sessionmap(this),
+    damage_table(whoami_, g_conf().get_val<bool>("mds_damage_log_to_file"),
```
One can use less bulky "g_conf->mds_damage_log_to_file" format here and there since the config parameters have "with legacy : true".

ifed01 commented on src/mds/DamageTable.h:
```diff
@@ -171,10 +187,36 @@ class DamageTable
+    void set_log_to_file(bool _log_to_file) {
+      log_to_file = _log_to_file;
+      if (log_to_file) {
+        open_damage_log_file(fout, log_file);
```
May be bettet to check for non_empty log_file before calling open_damage_log_file() to avoid redundant errors on empy file opening attempt in the log.

ifed01 commented on src/mds/CInode.cc:
```diff
@@ -5231,23 +5230,45 @@ void CInode::scrub_initialize(ScrubHeaderRef& header)
-
   scrub_infop->scrub_in_progress = false;
   scrub_infop->header->dec_num_pending();
+  scrub_infop->remote_links.clear();
```
Perhaps need to set forward_scrub to true to have consistent inode state after abort too.

ifed01 commented on src/mds/ScrubStack.cc:
```diff
@@ -537,7 +682,33 @@ void ScrubStack::_validate_inode_done(CInode *in, int r,
+  in->scrub_reset_remote_links();
+
+  if (in->scrub_info()->forward_scrub) {
+    _enqueue(in, header, true);
```
I don't understand why do we need to enqueue it once again here - IIUC this will effectively do nothing as remote_links are emty and in->scrub_is_in_progress() == true

ifed01 replied:
And if this block is removed then scrub_reset_remote_links() call above is redundant as well.

sajibreadd-croit replied:
First of all scrubbing operation are not parallel in mds side, it parallalize the I/O towards osd only. 
```
void *MDSRank::ProgressThread::entry()
{
  std::unique_lock l(mds->mds_lock);
  while (true) {
    cond.wait(l, [this] {
      return (mds->stopping ||
	      !mds->finished_queue.empty() ||
	      (!mds->waiting_for_nolaggy.empty() && !mds->beacon.is_laggy()));
    });

    if (mds->stopping) {
      break;
    }

    mds->_advance_queues();
  }

  return NULL;
}
```
This progress thread runs every task sequentially and under big mds_lock
> I don't understand why do we need to enqueue it once again here

Let's say forward_scrub is false, in that case it will scrub backwards. Now for a specific inode it sends request to osd through `in->validate_disk_state(&fin->result, fin);`. This I/O is asynchronous. So in this async I/O, what if in between some directory recursed through that inode and tries to enqeue it for forward scrubbing. But as scrub is in progress (asynchronously) it won't push that inode into the queue but make forward_scrub = true here.
```
int ScrubStack::_enqueue(
    MDSCacheObject *obj, ScrubHeaderRef &header, bool top, bool *added,
    std::vector<std::pair<std::string, inodeno_t>> &&remote_links) {
  ceph_assert(ceph_mutex_is_locked_by_me(mdcache->mds->mds_lock));
  if (CInode *in = dynamic_cast<CInode*>(obj)) {
    if (in->scrub_is_in_progress()) {
      dout(10) << __func__ << " with {" << *in << "}" << ", already in scrubbing" << dendl;
      if (!remote_links.empty()) {
        in->scrub_add_remote_link(std::move(remote_links));
      } else {
        in->set_forward_scrub(true);
      }
      return -CEPHFS_EBUSY;
    }
```

sajibreadd-croit replied:
You are right. We need to make scrub_is_in_progress false before _enqueue call

ifed01 commented on src/mds/ScrubStack.cc:
```diff
@@ -408,12 +429,110 @@ class C_InodeValidated : public MDSInternalContext
+        path += dn->get_name();
+      }
+      stack->add_remote_link_damage(path, ino);
+      header->inc_scrubbed_remote_link_count();
```
This could be incremented within add_remote_link_damage

ifed01 commented on src/mds/ScrubHeader.h:
```diff
@@ -64,6 +64,18 @@ class ScrubHeader {
   }
   unsigned get_num_pending() const { return num_pending; }
 
+  void inc_scrubbed_inode_count() { ++scrubbed_inode_count; }
```
redundant

sajibreadd-croit replied:
Didn't understand what part of it is redundant?

ifed01 replied:
Doesn't matter, please ignoer

ifed01 commented on src/mds/ScrubStack.cc:
```diff
@@ -537,7 +680,34 @@ void ScrubStack::_validate_inode_done(CInode *in, int r,
+  }
+  in->scrub_reset_remote_links();
+
+  if (in->scrub_info()->forward_scrub && in->is_dir()) {
```
IMO more elegant would be:
bool do_enqueue = in->scrub_info()->forward_scrub && in->is_dir();
in->scrub_finished();
if(do_enqueue) {
  _enqueue(..);
}

ifed01 replied:
Reworked in https://github.com/croit/ceph/pull/7

ifed01 commented on src/mds/ScrubStack.cc:
```diff
@@ -408,12 +429,109 @@ class C_InodeValidated : public MDSInternalContext
+                    {std::make_pair(std::move(path), ino)});
+    stack->kick_off_scrubs();
+  safe_exit:
+    dn->put(MDSCacheObject::PIN_SCRUBQUEUE);
```
IMO this code can be placed in dtor (which would ideally match the corresponding actions performed in ctor) and hence get rid off goto statements in this func

ifed01 replied:
fixed in (https://github.com/croit/ceph/pull/7)

ifed01 commented on src/mds/ScrubStack.cc:
```diff
@@ -408,12 +429,109 @@ class C_InodeValidated : public MDSInternalContext
+      path += dn->get_name();
+      mdcache->mds->damage_table.notify_remote_link_damaged(
+          dnl->get_remote_ino(), path);
+      return nullptr;
```
Shouldn't we call  
header->inc_scrubbed_remote_link_count();

before returning here?

ifed01 replied:
(fixed in https://github.com/croit/ceph/pull/7)

ifed01 commented on src/mds/ScrubStack.cc:
```diff
@@ -454,11 +572,19 @@ void ScrubStack::scrub_dirfrag(CDir *dir, bool *added_children, bool *done)
+        CInode *remote_inode = remote_link_checkup(dn, header);
+        if (remote_inode && !header->get_repair() &&
+            g_conf()->mds_force_hard_link_scrubbing) {
+          std::string remote_path;
```
And here is one more remote_path calculation - which really resembles what's performed within remote_link_checkup(). Couldn't we have a single point for this path creation?

ifed01 replied:
Reworked in https://github.com/croit/ceph/pull/7

ifed01 commented on src/mds/ScrubStack.cc:
```diff
@@ -75,6 +81,11 @@ int ScrubStack::_enqueue(MDSCacheObject *obj, ScrubHeaderRef& header, bool top)
 
     dout(10) << __func__ << " with {" << *in << "}" << ", top=" << top << dendl;
     in->scrub_initialize(header);
+    if (!remote_links.empty()) {
```
Better move this stuff into scrub_initialize

ifed01 replied:
fixed in (https://github.com/croit/ceph/pull/7)

ifed01 commented on src/mds/ScrubStack.cc:
```diff
@@ -537,7 +680,34 @@ void ScrubStack::_validate_inode_done(CInode *in, int r,
+      }
+    }
+  }
+  in->scrub_reset_remote_links();
```
redundant, scrub_finished() does the same a few lines below

ifed01 replied:
Fixed in https://github.com/croit/ceph/pull/7

ifed01 commented on src/mds/ScrubStack.cc:
```diff
@@ -408,12 +429,109 @@ class C_InodeValidated : public MDSInternalContext
+    }
+    stack->_enqueue(remote_inode, header, true, nullptr,
+                    {std::make_pair(std::move(path), ino)});
+    stack->kick_off_scrubs();
```
We probably should call this no matter if damage has been detected or not....

ifed01 commented on src/mds/ScrubStack.cc:
```diff
@@ -408,12 +429,109 @@ class C_InodeValidated : public MDSInternalContext
+  MDCache* mdcache;
+  C_RemoteInodeOpened(MDSRank *mds, ScrubStack *stack_,
+                       ScrubHeaderRef &header_, CDentry *dn_, inodeno_t ino_)
+      : MDSInternalContext(mds), stack(stack_), header(header_), dn(dn_),
```
header/dn initialization order warning

ifed01 replied:
fixed in (https://github.com/croit/ceph/pull/7)

ifed01 committed:
mds: refactor hard link scrubbing

Signed-off-by: Igor Fedotov <igor.fedotov@croit.io>

sajibreadd-croit committed:
Introduce end-to-end targeted `scan_extents` support by wiring inode/damage-file
candidate selection, type filtering, and bounded window controls
(`--extent-period`, `--extent-limit`) into DataScan while preserving legacy
full-scan behavior. Add cls-assisted accumulated-metadata flushing semantics
and robust CephFS QA fixtures/tests that validate manual expectations,
inode/damage parity, multi-pool edge cases, and argument validation with
deterministic QA coverage.
Signed-off-by: Md Mahamudur Rahaman Sajib <mahamudur.sajib@croit.io>

sajibreadd-croit committed:
Added targeted scanning for scan_inodes
Signed-off-by: Md Mahamudur Rahaman Sajib <mahamudur.sajib@croit.io>

sajibreadd-croit committed:
removed inodes should be removed from injected_inos as well.
Signed-off-by: Md Mahamudur Rahaman Sajib <mahamudur.sajib@croit.io>

sajibreadd-croit committed:
Parallelized scan_links
Signed-off-by: Md Mahamudur Rahaman Sajib <mahamudur.sajib@croit.io>

ifed01 commented on src/tools/cephfs/DataScan.h:
```diff
@@ -296,11 +352,29 @@ class DataScan : public MDSUtility, public MetadataTool
+    std::string damage_file_path;
+    std::string damage_type_expr;
+    std::vector<std::string> damage_type_tokens;
+    bool damage_type_filter_set;
```
Looks redundant, could be replaced with damage_type_tocken.size()

ifed01 commented on src/tools/cephfs/DataScan.h:
```diff
@@ -296,11 +352,29 @@ class DataScan : public MDSUtility, public MetadataTool
+    std::string damage_type_expr;
+    std::vector<std::string> damage_type_tokens;
+    bool damage_type_filter_set;
+    bool damage_type_all;
```
Isn't that always equal to !damage_type_filter_set ?

ifed01 commented on src/tools/cephfs/DataScan.h:
```diff
@@ -296,11 +352,29 @@ class DataScan : public MDSUtility, public MetadataTool
+    // Parser state for scan_extents damage/inode inputs.
+    std::string damage_file_path;
+    std::string damage_type_expr;
+    std::vector<std::string> damage_type_tokens;
```
why not using std::set instead?

ifed01 commented on src/tools/cephfs/DataScan.h:
```diff
@@ -296,11 +352,29 @@ class DataScan : public MDSUtility, public MetadataTool
+    bool damage_type_all;
+    std::string inode_file_path;
+    uint64_t extent_period;
+    bool extent_period_set;
```
IMO redundant, extent_period == 0 should be sufficient to indicate unset val

sajibreadd-croit replied:
I am keeping the default value of `extent_period = 1` such that if someone doesn't add that parameter then default value will be kept. But also I want to protect other commands having extent_period. That's why a sanity boolean is used.

ifed01 commented on src/tools/cephfs/DataScan.h:
```diff
@@ -296,11 +352,29 @@ class DataScan : public MDSUtility, public MetadataTool
+    uint64_t extent_period;
+    bool extent_period_set;
+    uint64_t extent_limit;
+    bool extent_limit_set;
```
same here

ifed01 commented on src/tools/cephfs/DataScan.cc:
```diff
@@ -631,6 +1164,356 @@ int DataScan::scan_extents()
   return 0;
 }
 
+int DataScan::scan_extents_for_damage_file()
```
IMO better to refactor parsing code in a way that damage file and ino file parsings aren't copy-pasted.
What's about having two methods to parse these files which get function pointer as a parameter to call for specific processing?

ifed01 commented on src/tools/cephfs/DataScan.cc:
```diff
@@ -568,6 +1011,96 @@ int parse_oid(const std::string &oid, uint64_t *inode_no, uint64_t *obj_id)
+    bool all_stat_failed = true;
+
+    uint64_t window_end = window_start + extent_period;
+    if (window_end < window_start || window_end > extent_limit) {
```
How can window_end be less than window_start? IMO that's a redundant check...

ifed01 commented on src/cls/cephfs/cls_cephfs_client.cc:
```diff
@@ -27,21 +27,50 @@ using ceph::decode;
+    librados::IoCtx &ctx, inodeno_t inode_no, const uint64_t obj_index,
+    const uint64_t obj_size, const int64_t obj_pool_id, const time_t mtime,
+    AccumulateResult *accum_res, const bool immediate_write) {
+  if (!immediate_write) {
```
Why are you extending accumulate_inode_metadata() interface with two additional parameter to perform completely different in-mem accumulation instead of having/using standalone method, e.g. AccumulatedResult::update_accumulate_result or something.

ifed01 commented on src/cls/cephfs/cls_cephfs_client.cc:
```diff
@@ -27,21 +27,50 @@ using ceph::decode;
+    const uint64_t obj_size, const int64_t obj_pool_id, const time_t mtime,
+    AccumulateResult *accum_res, const bool immediate_write) {
+  if (!immediate_write) {
+    if (accum_res == nullptr) {
```
IMO one can assert here and get rid off any return value

ifed01 commented on src/tools/cephfs/DataScan.cc:
```diff
@@ -568,6 +1011,96 @@ int parse_oid(const std::string &oid, uint64_t *inode_no, uint64_t *obj_id)
+        r = ClsCephFSClient::accumulate_inode_metadata(data_io, ino, extent_id,
+                                                       size, obj_pool_id, mtime,
+                                                       &accum_res, false);
+        if (r < 0) {
```
One can get rid off any error handling here if using standalone update_accumulate_result() method for in-mem update as per my comment above.

ifed01 commented on src/cls/cephfs/cls_cephfs_client.cc:
```diff
@@ -62,6 +91,24 @@ int ClsCephFSClient::accumulate_inode_metadata(
   return ctx.operate(zeroth_object.name, &op);
 }
 
+int ClsCephFSClient::flush_inode_accumulate_result(
```
I have a feeling that everything this function does could be performed through the original accumulate_inode_metadata() call.  Am I missing something? 
If thats' the case then all these new code to support set_accumulated_inode_metadata method and its parameter structure could be removed.

sajibreadd-croit replied:
Actually it won't be possible (I also initially think that way). Let's assume a case where no chunk exist but the hard_link (basically as dentry) exist. In that case I would like to create a fake inode using `force_create_head_inode` to resolve the issue. We could remove that dentry from omap as well but it would cause extra corruption in link count. Those fake inode eventually will go into directory `lost+found` (check `scan_inode_from_oid` function).

ifed01 commented on src/tools/cephfs/DataScan.cc:
```diff
@@ -568,6 +1011,96 @@ int parse_oid(const std::string &oid, uint64_t *inode_no, uint64_t *obj_id)
+      }
+    }
+
+    if (!window_productive && all_stat_failed) {
```
Isn't window_production check redundant here? IIUC when all_stat_failed is true window_productive is always false.

ifed01 commented on src/tools/cephfs/DataScan.cc:
```diff
@@ -1970,11 +2207,23 @@ int DataScan::scan_frags()
+                                 uint64_t obj_name_offset,
+                                 bool force_restore_ancestors,
+                                 std::unordered_set<uint64_t> &&inode_set) {
+  if (force_restore_ancestors &&
```
do we really need this check here or we can rely on the same inside inject_with_backtrace?

sajibreadd-croit replied:
Did you mean the cycle check is unnecessary or the whole `scan_frag_from_oid` call inside `inject_with_backtrace` is unnecessary?

ifed01 commented on src/tools/cephfs/DataScan.cc:
```diff
@@ -2567,7 +2818,13 @@ int MetadataDriver::inject_with_backtrace(
       // should be fixed up using a separate tool scanning the metadata
       // pool.
-      break;
+      if (force_inject_ancestors) {
```
Is the comment above still valid with the new code?

sajibreadd-croit replied:
This will not be true according to the code I did. When I found a dir_frag, I used that directory inode object's backtrace to restore not the outdated backtrace of child inode.

sajibreadd-croit replied:
My argument here is, if this idea have risk then whole scan_frags is in risk.

sajibreadd-croit committed:
Review resolved (Will squash this commit later)

Signed-off-by: Md Mahamudur Rahaman Sajib mahamudur.sajib@croit.io
